### ISSUE TITLE
Infer Numba Types from Python Type Hints

### ISSUE THREAD

**Issue by donielix** [numba/numba#9475]
### Feature description

Currently, Numba requires explicit specification of function signatures in the `njit` decorator, which can be verbose and redundant, especially when Python type hints are already provided. This feature request proposes the enhancement of the Numba `njit` decorator to automatically infer Numba types from Python type hints.

### Use case

This feature would streamline the development process by reducing the need for duplicate type information and leveraging Python's type hints for just-in-time (JIT) compilation with Numba. By using the `inspect` module or similar mechanisms to extract type hints from function definitions, Numba can automatically generate the appropriate JIT signatures, making the code cleaner and more Pythonic.

### Example

Currently, to use the `njit` decorator, we must explicitly define the types, as shown below:

```python
import numba as nb
@nb.njit("f8[:](f8[:,:], i8)")
def test(x, a):
    return x[0, :]**2 + a
```

With the proposed feature, developers could write more intuitive and readable code by leveraging Python type hints:

```python
from inspect import signature
import numpy as np

def new_njit(func):
    print(signature(func))
    def wrapper(*args, **kwargs):
        return func(*args, **kwargs)
    return wrapper

@new_njit
def test(x: np.ndarray[:,:], a: np.int64):
    return x[0, :]**2 + a
```

At compile time, this would output the type hints `(x: numpy.ndarray[slice(None, None, None), slice(None, None, None)], a: numpy.int64)`, which could be automatically translated into Numba types for JIT compilation.

### Benefits

- **Simplicity and Readability**: Eliminates the redundancy of specifying types both in Python type hints and Numba signatures, leading to cleaner code.
- **Ease of Use**: Makes it easier for developers familiar with Python type hints to use Numba without the steep learning curve of understanding Numba's type system.
- **Future-Proof**: As Python's type hinting system continues to evolve and gain adoption, this feature ensures that Numba remains compatible and takes full advantage of Python's native capabilities.

**Comment by jorenham** [comment]
Unlike the annotations in your example, there currently are valid ways for annotating the shape of an array. For instance, in Python 3.12 you can do:

```python
import numpy as np

type Array[Shape: tuple[int, ...], T: np.generic] = np.ndarray[Shape, np.dtype[T]]
type Vector[I: int, T: np.generic] = Array[tuple[I], T]
type Matrix[I: int, J: int, T: np.generic] = Array[tuple[I, J], T]

def outer[I, J, T](xi: Vector[I, T], xj: Vector[J, T]) -> Matrix[I, J, T]:
    ...
```

However, there appear to be no plans to implement something like this in numpy at the moment, see https://github.com/numpy/numpy/issues/16544 .
There are also several limitations to python typing that limit the ways in which this can be used, e.g. a `TypeVarTuple` cannot be bounded, and there are no higher kinded types (https://github.com/python/typing/issues/548).

---

A good example of how type hints can be used at runtime, is [Typer](https://typer.tiangolo.com/). They use `typing.Annotated` to piggyback typer-specific metadata on the annotations of function parameters. Type checkers don't mind this. 

So if you were to implement something like typer has done in numba, I can't imagine it being good for readability. Just imagine your example, but with all parameter type, and the return type, wrapped in a `typing.Annotated[{original_type}, {numba_spec}]`...

There's also the problem that figuring out the annotations at runtime is really difficult. Consider e.g. `from __future__ import annotation` that stringifies every annotation, or the new `type ...` alias, which wraps the alias into another `typing.TypeAliasType` so that even `typing.get_args()` won't work on it. And don't forget about generic functions. Implementing those would probably be an acceptable comp-sci MSc research subject.

So this idea would probably work on for some toy examples, for which you could hack together a propotype. But if you want to be serious about this, and really make it "future-proof", then you're going to have to invest at least a year of full-time research work on this. 

---

TLDR: It's not a bad idea, but given the current state of typing in python, I don't think it's very feasible if you want to do this right. You should at least wait until Numpy makes a decision on how to annotate the shapes of arrays.

**Comment by esc** [comment]
Expecting this to be a Legendary Quest with an insane difficulty rating... 🐉 ⚔️ 💰

**Comment by jorenham** [comment]
FYI, I made an attempt at a proposal for the typing of numpy array shapes:
https://github.com/numpy/numpy/issues/16544#issuecomment-2005169857
